### PR TITLE
[New] Navigate route

### PR DESCRIPTION
**Pull request by dfeinzimer**
## Description

This PR implements `Navigate route`. It was originally written by Christopher but not fully completed/merged.
[README](https://github.com/Esri/arcgis-maps-sdk-swift-samples/tree/df/New-NavigateARoute/Shared/Samples/Navigate%20a%20route) 

## Linked Issue(s)

- `swift/issues/2254`

## How To Test

- 1: Tap "Navigate"
- 2: Optionally pan the map away and press "Recenter" to recenter on the current location
- 3: Optionally press "Reset" to stop navigation

## Screenshots

navigate-route

### CODE REVIEW

**Comment by mhdostal** [comment]
Folder/Group needs to be put in the sorted order.
image

**Comment by mhdostal** [comment]
The app crashes when tapping the "info" button in the "Navigate a route" sample; the `NavigateARouteView.swift` file needs to be added to the "Copy Source Code Files" section of the project.

File: Shared/Samples/Navigate a route/NavigateARouteView.swift
Comment (mhdostal) [inline Shared/Samples/Navigate a route/NavigateARouteView.swift:91]: This is never used by the view. Is it still coming?

**Review by mhdostal — CHANGES_REQUESTED** [review]
Just the one question and the conversations about the crash and folder order.

**Review by mhdostal — APPROVED** [review]

File: Samples.xcodeproj/project.pbxproj
Comment (yo1995) [inline Samples.xcodeproj/project.pbxproj:60]: Sorry for the late notice. Can you help rename the sample to "Navigate route" according to the common design? There was an effort to standardize the names across platforms, and the old iOS sample name deviates from the common design. Thanks.

- the folder name
- README title
- title in metadata
- swift filename
- view type name (NavigateARouteView -> NavigateRouteView)

File: Shared/Samples/Navigate route/NavigateRouteView.swift
Comment (yo1995) [inline Shared/Samples/Navigate route/NavigateRouteView.swift:244]:
```suggestion
            viewpoint = Viewpoint(boundingGeometry: routeGeometry, rotation: 0)
```
Like to see the map rotate back to the initial state.

File: Shared/Samples/Navigate route/NavigateRouteView.swift
Comment (yo1995) [inline Shared/Samples/Navigate route/NavigateRouteView.swift:247]:
```suggestion
        /// Starts monitoring multiple asynchronous streams of information.
```

File: Shared/Samples/Navigate route/NavigateRouteView.swift
Comment (yo1995) [inline Shared/Samples/Navigate route/NavigateRouteView.swift:261]: I attempted to remove the `try` keyword and the code still works. 🤔  Out of curiosity, can you find out why/when the tracking status can be `nil`?

In iOS API, the delegate method `func routeTracker(_ routeTracker: AGSRouteTracker, didUpdate trackingStatus: AGSTrackingStatus)` doesn't give a `nil` status.

**Review by mhdostal — DISMISSED** [review]

File: Shared/Samples/Navigate route/NavigateRouteView.swift
Comment (yo1995) [inline Shared/Samples/Navigate route/NavigateRouteView.swift:270]:
```suggestion
                        Time remaining: \(timeFormatter.string(from: status.routeProgress.remainingTime)!)
```
From the doc I think it is safe to force unwrap.

File: Shared/Samples/Navigate route/NavigateRouteView.swift
Comment (yo1995) [inline Shared/Samples/Navigate route/NavigateRouteView.swift:281]: I prefer not to reset the navigation to the start when it ends. Rather, I'd leave it to the user to tap on the reset button.

File: Shared/Samples/Navigate route/NavigateRouteView.swift
Comment (yo1995) [inline Shared/Samples/Navigate route/NavigateRouteView.swift:256]:
```suggestion
        /// Monitors the asynchronous stream of tracking statuses.
```

File: Shared/Samples/Navigate route/NavigateRouteView.swift
Comment (yo1995) [inline Shared/Samples/Navigate route/NavigateRouteView.swift:62]: Seems to me that `model.autoPanMode` only changes when the navigation is ongoing.
```swift
.disabled(model.autoPanMode == .navigation)
```
If you don't set the `model.autoPanMode` to `.off` in the initializer, but to `.navigation` either in the initializer or directly on the property's declaration, then I think this condition can be simplified.

It is OK if you don't think it can be removed.

File: Shared/Samples/Navigate route/NavigateRouteView.swift
Comment (yo1995) [inline Shared/Samples/Navigate route/NavigateRouteView.swift:323]: If the suggestion above is applied, then this line can be removed.

**Review by yo1995 — COMMENTED** [review]
First pass comments. I'll take a closer look at the booleans once some other suggestions are addressed.

File: Shared/Samples/Navigate route/NavigateRouteView.swift
Comment (dfeinzimer) [reply Shared/Samples/Navigate route/NavigateRouteView.swift:62]: `Model.autoPanMode` can be initialized to `.navigation` but the button will still be enabled then when there's no location to snap to. I think the condition should remain.

File: Shared/Samples/Navigate route/NavigateRouteView.swift
Comment (dfeinzimer) [reply Shared/Samples/Navigate route/NavigateRouteView.swift:208]: swift/issues/3842

**Comment by dfeinzimer** [comment]
Conflicts resolved, feedback applied - this is ready for review again.

File: Shared/Samples/Navigate route/NavigateRouteView.swift
Comment (yo1995) [reply Shared/Samples/Navigate route/NavigateRouteView.swift]: Hmm interesting. I see Kotlin generates the same code (navigation/RouteTracker.kt#L124), while in cocoa it only passes non optional status. If the status is `nil` only for the first empty location, then I think it's worth logging an issue to improve this API.

**Review by mhdostal — DISMISSED** [review]

File: Shared/Samples/Navigate route/NavigateRouteView.swift
Comment (yo1995) [reply Shared/Samples/Navigate route/NavigateRouteView.swift]: After discussion, it is an expected behavior.

**Comment by yo1995** [comment]
Initially I thought `isNavigateDisabled` and `isNavigatingRoute` can be combined together as 1 bool to control the various states of the buttons. But I found that combined bool will take too much responsibility and is vague in meaning, so I like what you have here.

File: Shared/Samples/Navigate route/NavigateRouteView.swift
Comment (dfeinzimer) [reply Shared/Samples/Navigate route/NavigateRouteView.swift:66]: I think it may have something to do with the viewpoint being set in WGS84 from the route geometry and it coming back out at as web mercator in `onViewpointChanged` but I'd have to debug further. Is it a blocker for this sample?

Comment (dfeinzimer) [reply Shared/Samples/Navigate route/NavigateRouteView.swift:66]: I dug some more; on line 248 the viewpoint is created without a scale (`viewpoint = Viewpoint(boundingGeometry: routeGeometry, rotation: 0)`. If a scale is passed, (e.g. `viewpoint = Viewpoint(center: routeGeometry.extent.center, scale: 34182, rotation: 0)`) then the provided scale is preserved when `onViewpointChanged` fires.

Comment (yo1995) [reply Shared/Samples/Navigate route/NavigateRouteView.swift:66]: It doesn't block this sample. I think it might be worth logging an issue to take a further look though.

Comment (dfeinzimer) [reply Shared/Samples/Navigate route/NavigateRouteView.swift:66]: `swift/issues/3854`

**Review by yo1995 — COMMENTED** [review]
Will approve after the merge conflict is solved.

**Comment by dfeinzimer** [comment]
Feedback applied, conflicts resolved

**Review by mhdostal — APPROVED** [review]

File: Shared/Samples/Navigate route/NavigateRouteView.swift
Comment (yo1995) [inline Shared/Samples/Navigate route/NavigateRouteView.swift:138]: This formatter is still needed for iOS 15: https://stackoverflow.com/a/72945284/14369688

**Pull request merged**
